### PR TITLE
[CELEBORN-919] Move Columnar Shuffle code into an individual module

### PR DESCRIPTION
**Pull request by zhouyifan279**
### What changes were proposed in this pull request?

Move Columnar Shuffle code into an individual module

### Why are the changes needed?

Spark 3.5 made a lot of changes to AtomicType in https://issues.apache.org/jira/browse/SPARK-42887.

This causes compilation error when building columnar shuffle code.

As columnar shuffle is a configurable feature, I think it's better to move related code into a individual module. Then we can exclude this module when build with Spark 3.5 for now.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?

Add test `ColumnarHashBasedShuffleWriterSuiteJ` and `CelebornColumnarShuffleReaderSuite`

### CODE REVIEW

**Comment by zhouyifan279** [comment]
cc @cfmcgrady @pan3793

File: client-spark/spark-3-columnar-shuffle/src/main/java/org/apache/spark/shuffle/celeborn/CustomShuffleDependencyUtils.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
+package org.apache.spark.shuffle.celeborn;
```
Comment (zwangsheng) [inline client-spark/spark-3-columnar-shuffle/src/main/java/org/apache/spark/shuffle/celeborn/CustomShuffleDependencyUtils.java:1]: License header

File: client-spark/spark-3-columnar-shuffle/src/test/scala/org/apache/spark/shuffle/celeborn/CelebornColumnarShuffleReaderSuite.scala [hunk 2]
```diff
@@ -0,0 +1,42 @@
+package org.apache.spark.shuffle.celeborn
```
Comment (zwangsheng) [inline client-spark/spark-3-columnar-shuffle/src/test/scala/org/apache/spark/shuffle/celeborn/CelebornColumnarShuffleReaderSuite.scala:1]: ditto

File: client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala [hunk 3]
```diff
@@ -50,21 +49,7 @@ class CelebornShuffleReader[K, C](
...
-          dep.serializer.asInstanceOf[UnsafeRowSerializer])
-        serializerInstance = new CelebornColumnarBatchSerializer(
-          schema,
-          conf.columnarShuffleBatchSize,
-          conf.columnarShuffleDictionaryEnabled,
-          conf.columnarShuffleOffHeapEnabled,
-          dataSize).newInstance()
-      }
-    }
+    val serializerInstance = newSerializerInstance(dep)
```
Comment (zwangsheng) [inline client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:52]: Consider we merged #1838, we should rebase and tweak the logic below:

```scala
    val iterWithUpdatedRecordsRead =
      if (GlutenColumnarBatchSerdeHelper.isGlutenSerde(serializerInstance.getClass.getName)) {
        GlutenColumnarBatchSerdeHelper.withUpdatedRecordsRead(recordIter, metrics)
      } else {
        recordIter.map { record =>
          metrics.incRecordsRead(1)
          record
        }
      }
```

**Comment by waitinfuture** [comment]
cc @kerwin-zk

File: client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala (hunk 3 above)
Comment (zhouyifan279) [reply client-spark/spark-3/src/main/scala/org/apache/spark/shuffle/celeborn/CelebornShuffleReader.scala:52]: Gluten Columnar Shuffle is different from Celeborn Columnar Shuffle.
It is clean and won't stop us to upgrade spark dependency version.
Let's leave it here for now.

**Comment by kerwin-zk** [comment]
LGTM. Thanks!

**Review by kerwin-zk — APPROVED** [review]

**Comment by pan3793** [comment]
Thanks, merged to main

File: project/CelebornBuild.scala [hunk 4]
```diff
@@ -540,7 +576,20 @@ trait SparkClientProjects {
         ) ++ commonUnitTestDependencies
       )
   }
-  
+
+  def sparkColumnarShuffle: Project = {
```
Comment (cfmcgrady) [inline project/CelebornBuild.scala:580]: let's add the project `sparkColumnarShuffle` to the spark tests group `sparkGroup` to enable the columnar-related tests for SBT.

File: client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java [hunk 5]
```diff
@@ -304,17 +309,8 @@ public <K, C> ShuffleReader<K, C> getReaderForRange(
...
-          h,
-          startPartition,
-          endPartition,
-          startMapIndex,
-          endMapIndex,
-          context,
-          celebornConf,
-          metrics);
+      return getCelebornShuffleReader(
+          handle, startPartition, endPartition, startMapIndex, endMapIndex, context, metrics);
```
[the remark this answers is not in the document]
Comment (zhouyifan279) [reply client-spark/spark-3/src/main/java/org/apache/spark/shuffle/celeborn/SparkShuffleManager.java:313]: I created a new PR #1853 to unify the order.

File: project/CelebornBuild.scala (hunk 4 above)
Comment (zhouyifan279) [reply project/CelebornBuild.scala:580]: Created a new PR #1854

**Pull request closed without merging**